decay.go

```diff
@@ -0,0 +1,268 @@
+	// use real time if the Clock in the config is nil.
+	if cfg.Clock == nil {
+		cfg.Clock = clock.New()
+	}
+
+	d := &decayer{
+		cfg:       cfg,
+		mgr:       mgr,
+		clock:     cfg.Clock,
+		knownTags: make(map[string]*decayingTag),
+		bumpCh:    make(chan bumpCmd, 128),
+		closeCh:   make(chan struct{}),
+		doneCh:    make(chan struct{}),
+	}
+
+	// kick things off.
+	go d.process()
+
+	return d, nil
+}
+
+// decayingTag represents a decaying tag, with an associated decay interval, a
+// decay function, and a bump function.
+type decayingTag struct {
+	trkr      *decayer
+	name      string
+	interval  time.Duration
+	nextRound uint64
+	decayFn   connmgr.DecayFn
+	bumpFn    connmgr.BumpFn
+}
+
+func (d *decayer) RegisterDecayingTag(name string, interval time.Duration, decayFn connmgr.DecayFn, bumpFn connmgr.BumpFn) (connmgr.DecayingTag, error) {
+	d.tagsMu.Lock()
+	defer d.tagsMu.Unlock()
+
+	tag, ok := d.knownTags[name]
+	if ok {
+		return nil, fmt.Errorf("decaying tag with name %s already exists", name)
+	}
+
+	if interval < d.cfg.Resolution {
+		err := fmt.Errorf("decay interval for %s (%s) is lower than tracker's resolution (%s)",
+			name,
+			interval,
+			d.cfg.Resolution)
+		return nil, err
+	}
+
+	if interval%d.cfg.Resolution != 0 {
+		err := fmt.Errorf("decay interval for tag %s (%s) is not a multiple of tracker's resolution (%s); "+
+			"preventing undesired loss of precision",
+			name,
+			interval,
+			d.cfg.Resolution)
+		return nil, err
+	}
+
+	initRound := atomic.LoadUint64(&d.currRound) + uint64(interval/d.cfg.Resolution)
+	tag = &decayingTag{
+		trkr:      d,
+		name:      name,
+		interval:  interval,
+		nextRound: initRound,
+		decayFn:   decayFn,
+		bumpFn:    bumpFn,
+	}
+
+	d.knownTags[name] = tag
+	return tag, nil
+}
+
+// Close closes the Decayer. It is idempotent.
+func (d *decayer) Close() error {
+	select {
+	case <-d.doneCh:
+		return d.err
+	default:
+	}
+
+	close(d.closeCh)
+	<-d.doneCh
+	return d.err
+}
+
+// process is the heart of the tracker. It performs the following duties:
+//
+//  1. Manages decay.
+//  2. Applies score bumps.
+//  3. Yields when closed.
+func (d *decayer) process() {
+	defer close(d.doneCh)
+
+	ticker := d.clock.Ticker(d.cfg.Resolution)
+	defer ticker.Stop()
+
+	var (
+		bmp   bumpCmd
+		now   time.Time
+		visit = make(map[*decayingTag]struct{})
+	)
+
+	for {
+		select {
+		case now = <-ticker.C:
+			round := atomic.AddUint64(&d.currRound, 1)
+
+			for _, tag := range d.knownTags {
+				if tag.nextRound <= round {
+					// Mark the tag to be updated in this round.
+					visit[tag] = struct{}{}
+				}
+			}
+
+			// Visit each peer, and decay tags that need to be decayed.
+			for _, s := range d.mgr.segments {
+				s.Lock()
+
+				// Entered a segment that contains peers. Process each peer.
+				for _, p := range s.peers {
+					for tag, v := range p.decaying {
+						if _, ok := visit[tag]; !ok {
+							// skip this tag.
+							continue
```

Comment:
I would use named loops here to avoid confusion.

Reply:
We'd only name the inner loop, though, and the `for` is just two lines above the `continue`...

Reply:
ok, fair enough.